05_concurrency/homework/main.go

```diff
@@ -0,0 +1,89 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type Cache interface {
+	Set(key string, value interface{}, ttl time.Duration)
+	Get(key string) (interface{}, bool)
+	Delete(key string)
+}
+
+type CacheMap struct {
+	m       map[string]Value
+	counter int
+	mu      sync.Mutex
+}
+
+type Value struct {
+	value       interface{}
+	ttl         time.Duration
+	currentTime time.Time
+}
+
+func (t *CacheMap) Set(key string, value interface{}, ttl time.Duration) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	var valueFromMap = Value{value, ttl, time.Now()}
+	t.m[key] = valueFromMap
+	t.counter++
+	if t.counter > 1000000 {
+		t.clearCache()
+		t.counter = 0
+	}
+
+}
+
+func (t CacheMap) Get(key string) (interface{}, bool) {
```

Comment:
Method `Get` defined on value so mutex is being copied. You lock and unlock the copy of mutex but the original one. Same related to other methods.